Initial frappe accounting app

akhilnarang opened the pull request.

akhilnarang committed:
feat: Initialize App

akhilnarang committed:
chore: add in item doctype

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: add in warehouse doctype

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: add in stock ledger entry doctype

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: add in partial stock entry doctype

Need to figure out how to use items table here

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: add in stock entry table item doctype

Add in items table to stock entry using that

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: add in some basic tests

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: drop quantity from item doctype

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: update doctype name

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(stock_entry): drop child table attribute

Not sure how it got set here

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: add in source_warehouse, make both not required - we'll add validation in code

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: add in rate, make warehouse optional

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: allow ledger to have negative quantity

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat(stock_entry): add in some validation, make sure ledger entry is created as well

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore(stock_ledger_entry): add in rate

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore(stock_entry_table_item): make rate mandatory

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: add in stock balance report

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: import some configurations from frappe

- editorconfig
- black and isort configs

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: add in some tests

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on accounting/accounting/doctype/stock_entry/stock_entry.py:
```diff
@@ -0,0 +1,122 @@
+		).insert()
+
+	def before_save(self):
+		self.now = frappe.utils.now_datetime()
```
`now` is not a good variable name for this IMO. `posting_time` or `transaction_time` explain it better. 

We usually prefer having label and fieldname be as human-understandable as possible. label and fieldname are also usually similar, just `scrubbed`

akhilnarang replied:
5db8e55

ankush commented on accounting/accounting/doctype/stock_entry/stock_entry.py:
```diff
@@ -0,0 +1,122 @@
+			}
+		).insert()
+
+	def before_save(self):
```
This doctype should be submittable. Stock movement is kind of a "transaction" that can be 
- in draft when user is creating it and modifying it
- submitted - finalized by user and ledger impact is made
- cancelled - created by mistake, so document is cancelled and ledger impact is reversed.

ankush replied:
So you'll have to use two new hooks: on_submit and on_cancel.

ankush commented on accounting/accounting/doctype/stock_entry/stock_entry.py:
```diff
@@ -0,0 +1,122 @@
+		)
+
+		# Ensure that the warehouse has enough stock
+		if stock is None or item.quantity > stock:
```
```suggestion
		if item.quantity > cint(stock):
```

cint/cstr/flt etc are useful function avoid None checks. They also handle bunch of other things in mostly sane manner.

akhilnarang replied:
adf09f8

ankush commented on accounting/accounting/report/stock_balance/stock_balance.py:
```diff
@@ -0,0 +1,198 @@
+    )
+    for entry in item_warehouse_pairs:
+        # Get opening stock
+        if opening_stock := frappe.db.sql(
```
Rewrite all raw sql queries using ORM or QB
https://frappeframework.com/docs/user/en/api/query-builder

ankush commented on accounting/accounting/doctype/stock_entry/stock_entry.py:
```diff
@@ -0,0 +1,122 @@
+
+
+class StockEntry(Document):
+	def validate_item_metadata(self, item: StockEntryTableItem):
```
You can enable type exporting, "beta" feature rn. 

https://github.com/frappe/frappe/pull/21776

akhilnarang replied:
Hmm pretty nice, 73f448a

ankush commented on accounting/accounting/doctype/stock_entry/stock_entry.js:
```diff
@@ -0,0 +1,8 @@
+
+// frappe.ui.form.on("Stock Entry", {
+// 	refresh(frm) {
+
```
There should be some client side code to compute computed fields like total/amount on client side.

ankush commented on accounting/accounting/doctype/stock_entry/stock_entry.py:
```diff
@@ -0,0 +1,122 @@
+		match self.entry_type:
+			case "Receipt":
+				for item in self.items:
+					self.validate_item_metadata(item)
+					self.validate_receipt(item)
+					self.insert_ledger(item.item, item.target_warehouse, item.quantity, item.rate)
+			case "Consume":
+				for item in self.items:
+					self.validate_item_metadata(item)
+					self.validate_consume(item)
+					self.insert_ledger(item.item, item.source_warehouse, -item.quantity, item.rate)
+			case "Transfer":
+				for item in self.items:
+					self.validate_item_metadata(item)
+					self.validate_transfer(item)
+					self.insert_ledger(item.item, item.source_warehouse, -item.quantity, item.rate)
+					self.insert_ledger(item.item, item.target_warehouse, item.quantity, item.rate)
```
This kind of feels repetative.. any way you can reduce repetition?

ankush commented on accounting/accounting/doctype/stock_entry/test_stock_entry.py:
```diff
@@ -0,0 +1,164 @@
+class TestStockEntry(FrappeTestCase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.incoming_warehouse_name = "".join(random.choices(string.ascii_letters, k=10))
+        self.outgoing_warehouse_name = "".join(random.choices(string.ascii_letters, k=10))
+        self.main_warehouse_name = "".join(random.choices(string.ascii_letters, k=10))
```
There are some helper functions for this. 

frappe.generate_hash(), `random_string`

---

We should have some better API for faster fixture generation on FrappeTestCase.

ankush commented on .gitignore:
```diff
@@ -0,0 +1,7 @@
+*.swp
+tags
+node_modules
+__pycache__
```
Unrelated. This list is kind of small. 

Update this in Framework. We can create a mix of github's python + JS template.

ankush commented on accounting/accounting/doctype/item/test_item.py:
```diff
@@ -0,0 +1,48 @@
+
+
+class TestItem(FrappeTestCase):
+	def test_create_item_without_mandatory_fields(self):
```
Nice tests for app. 

Note for future: Doc creation, permission etc are considered framework features. So we kinda assume they work for the most part and don't add separate tests for them.

ankush commented on accounting/accounting/doctype/stock_entry/stock_entry.json:
```diff
@@ -0,0 +1,68 @@
+   "options": "Receipt\nConsume\nTransfer"
+  },
+  {
+   "fieldname": "target_warehouse",
+   "fieldtype": "Link",
+   "in_list_view": 1,
+   "label": "Target Warehouse",
+   "options": "Warehouse"
+  },
+  {
+   "fieldname": "items",
+   "fieldtype": "Table",
+   "label": "Items",
+   "options": "Stock Entry Table Item",
+   "reqd": 1
+  },
+  {
+   "fieldname": "source_warehouse",
+   "fieldtype": "Link",
+   "in_list_view": 1,
+   "label": "Source Warehouse",
+   "options": "Warehouse"
+  }
```
Add conditional mandatory here? source is mandatory for transfer and consume, target is mandatory for transfer and receipt. 

Check mandatory depends on on docfield.

ankush commented on accounting/accounting/doctype/stock_entry/stock_entry.py:
```diff
@@ -0,0 +1,122 @@
+
+class StockEntry(Document):
+	def validate_item_metadata(self, item: StockEntryTableItem):
+		if item.quantity < 1:
```
It can be decimal :grimacing:

akhilnarang replied:
Hmm changed it to a float value now, initial implementation had it as integer.

ankush commented on accounting/accounting/doctype/stock_entry/stock_entry.py:
```diff
@@ -0,0 +1,122 @@
+		)
+
+		# Ensure that the warehouse has enough stock
+		if stock is None or item.quantity > stock:
```
```suggestion
		if item.quantity > flt(stock):
```

ankush commented on accounting/accounting/doctype/stock_entry/stock_entry.py:
```diff
@@ -0,0 +1,122 @@
+			item.rate = average_rate
+
+	def insert_ledger(self, item: str, warehouse: str, quantity: int, rate: float):
+		frappe.get_doc(
```
```suggestion
		frappe.new_doc(
```


New doc is better when you're inserting a document:
- Semantically correct
- There are some edge cases with using get_doc over new_doc, so prefer new doc wherever possible

ankush commented on accounting/accounting/doctype/stock_entry/test_stock_entry.py:
```diff
@@ -0,0 +1,164 @@
+            }
+        )
+
+    def tearDown(self):
+        frappe.db.rollback()
```
FYI FrappeTestCase does this after running all tests (not individual tests)

ankush commented on accounting/accounting/doctype/stock_entry/test_stock_entry.py:
```diff
@@ -0,0 +1,164 @@
+
+        # Check ledger entries
+        for item in items:
+            data = frappe.db.get(
```
```suggestion
            data = frappe.db.get_value(
```


get is kinda deprecated.

ankush commented on accounting/accounting/doctype/stock_entry_table_item/stock_entry_table_item.json:
```diff
@@ -0,0 +1,66 @@
+{
+ "actions": [],
```
There seem to be two table doctypes... remove one? 

also naming should be `Stock Entry Item` as per our naming conventions

ankush commented on accounting/accounting/doctype/stock_ledger_entry/stock_ledger_entry.json:
```diff
@@ -0,0 +1,79 @@
+ "editable_grid": 1,
+ "engine": "InnoDB",
+ "field_order": [
+  "item",
+  "warehouse",
+  "entry_time",
+  "quantity",
+  "rate"
```
This should refer to the document that created it, else there's no clear linkages. 


Since more than one document type can create ledger, use "dynamic link".

ankush commented on accounting/accounting/report/stock_balance/stock_balance.py:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    if filters.item:
+        query_string += " AND item = %(item)s"
```
use query builder for this.

akhilnarang committed:
refactor: update and cleanup tests

Extract reusable code into separate methods

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(stock_entry): update variable name to make more sense

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(stock_entry): make use of frappe's `cint()` helper function

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore(hooks): enable experimental type annotation generation

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: regenerate doctypes with type annotations

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(stock_balance): make use for frappe's `flt()` helper function

Make it a float value instead of integer

Also return the absolute value of the outgoing stock, negative doesn't make sense since its already __outgoing__

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: drop usage of get_doc() where it doesn't seem to make sense

Use new_doc() for inserting documents, and db.get_value() for querying

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: rename stock entry table item to follow conventions

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: add in stock ledger report

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(stock_balance): use query builder

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore(test_stock_entry): drop teardown rollback

Frappe handles this already

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat(stock_ledger_entry): add in a dynamic link to depict the type/source of ledger entry

Make store entry submitable, only create ledger entry on submitting

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat(stock_entry): add `on_cancel()` hook to reverse a ledger entry when stock entry is changed

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat(stock_entry): set target and source warehouse as conditional mandatory

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(stock_entry): improve ledger creation code

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(test_stock_entry): submit the created document

Signed-off-by: Akhil Narang <me@akhilnarang.dev>
